File: source/pydgin.cpp
/*
Script for converting pdg to cpp, this time written in cpp.

Pretty much a line-for-line translation (ironically) of the .py.

Note: Should work, but my boost files aren't set up right lol
*/

#include <iostream>
#include <fstream>
#include "boost/regex.hpp"
#include <string>
#include <vector>
using namespace std;

vector<string> split(const string& text, const char& deliminator) {
    vector<string> vect;
    string temp;
    for (int i = 0; i < text.length(); i++) {
        if (text[i] == deliminator) {
            vect.push_back(temp);
            temp = "";
            continue;
        }
        temp += text[i];
    }
    return vect;
}

int main(int argc, char **argv) {
    // Takes in input filepath, output filepath
    if (argc < 3) {
        cout << "Invalid translator call.\n";
        return -1;
    }
    char* inp_filepath = argv[1];
    char* out_filepath = argv[2];

    // Get text from file
    ifstream file;
    file.open(inp_filepath, ios::in);
    string text, temp_s;
    text = "";
    while (getline(file, temp_s)) {
        text += temp_s;
    }
    file.close();

    // Determine if header
    bool header = text.substr(0, 7) == "@header";
    if (header) {
        text = text.substr(8);
    }

    // Semicolons
    text = boost::regex_replace(text, boost::regex("(?<![>:])\n"), ";\n");

    // Replace spaces with tabs
    text = boost::regex_replace(text, boost::regex("    "), "\t");

    // Tab pub/priv/restr
    text = boost::regex_replace(text, boost::regex("private"), "\\tprivate");
    text = boost::regex_replace(text, boost::regex("public"), "\\tpublic");
    text = boost::regex_replace(text, boost::regex("protected"), "\\tprotected");

    // Loop stuff
    text = boost::regex_replace(text, boost::regex("while (?=[^\\(])"), "while (");
    text = boost::regex_replace(text, boost::regex("for (?=[^\\(])"), "");
    text = boost::regex_replace(text, boost::regex("if (?=[^\\(])"), "");
    text = boost::regex_replace(text, boost::regex("else if (?=[^\\(])"), "");

    // Brackets, end parenthesis, for loop semicoloning
    vector<string> temp = split(text, '\n');
    for (int i = 0; i < temp.size(); i++) {
        temp[i] += '\n';
    }
    text = "";

    int tab_count_prev = 0;
    int tab_count_cur;
    string line, prefix;
    int opens;
    char character;
    for (int line_num = 0; line_num < temp.size(); line_num++) {
        line = temp[line_num];

        if (line.length() == 0) {
            continue;
        }

        // Fix for loop semicoloning
        if (boost::regex_search(line, boost::regex("(?<!\\w)for "))) {
            line = boost::regex_replace(line, boost::regex(","), ";");
        }

        // Balance loop parenthesis
        if (boost::regex_search(line, boost::regex("(for )|(while )|(if )|(else if )"))) {
            opens = 0;
            prefix = "";
            for (int i = 0; i < line.length(); i++) {
                character = line[i];
                if (character == '(') {
                    opens++;
                } else if (character == ')') {
                    opens--;
                }
                prefix += character;
                if (i + 1 == line.length() || line[i + 1] == ':') {
                    for (int j = 0; j < opens; j++) {
                        prefix += ')';
                    }
                    prefix += line.substr(i + 1);
                    break;
                }
            }
            temp[line_num] = prefix;
        }
        line = temp[line_num];

        tab_count_cur = 0;
        while (line[tab_count_cur] == '\t') {
            tab_count_cur++;
        }

        if (tab_count_cur > tab_count_prev) {
            for (int i = tab_count_prev; i < tab_count_cur; i++) {
                text = text.substr(0, text.size() - 2);
                text += " {\n";
            }
        } else if (tab_count_cur < tab_count_prev) {
            for (int i = tab_count_prev; i < tab_count_cur; i--) {
                for (int j = 0; j < i - 1; j++) {
                    text += '\t';
                }
                text += "}\n";
            }
        }

        text += line;
        tab_count_prev = tab_count_cur;
    }

    // Misc housekeeping
    text = boost::regex_replace(text, boost::regex("(?<=\\n).*#.*\\n"), "");
    text = boost::regex_replace(text, boost::regex("@import"), "#include");
    text = boost::regex_replace(text, boost::regex("(?<!(private)|(public)):(?!=[a-zA-Z0-9:])"), "");
    text = boost::regex_replace(text, boost::regex("['\"]{3}[^'\"]*['\"]{3}"), "");
    text = boost::regex_replace(text, boost::regex("\n"), "");
    text = boost::regex_replace(text, boost::regex(" in "), " : ");
    text = boost::regex_replace(text, boost::regex("\\); {"), ") {");
    text = boost::regex_replace(text, boost::regex("(?<=else);"), "");
    text = boost::regex_replace(text, boost::regex("@"), "#");

    // Fix classes
    temp = split(text, '\n');
    text = "";
    int to_semicolon = -1;
    int tabs;
    for (int i = 0; i < temp.size(); i++) {
        line = temp[i];
        tabs = 0;
        while (tabs + 1 < line.length() && line[tabs] == '\t') {
            tabs++;
        }
        if (boost::regex_search(line, boost::regex("(struct )|(class )"))) {
            to_semicolon = tabs;
        } else if (to_semicolon == tabs) {
            temp[i] += ';';
            to_semicolon = -1;
        }
    }
    for (int i = 0; i < temp.size(); i++) {
        text += temp[i] + '\n';
    }

    // Untab pub/priv/prot
    text = boost::regex_replace(text, boost::regex("\\tprivate"), "private");
    text = boost::regex_replace(text, boost::regex("\\tpublic"), "public");
    text = boost::regex_replace(text, boost::regex("\\tprotected"), "protected");

    // Trim trailing whitespace
    while (text[text.length() - 1] == '\n' || text[text.length() - 1] == ' ') {
        text = text.substr(0, text.length() - 1);
    }
    text += '\n';

    // Fix scope resolution operator
    text = boost::regex_replace(text, boost::regex(" get "), "::");

    // Write to output
    ofstream out_file;
    out_file.open(out_filepath, ios::out);
    out_file.seekp(ios::beg);
    out_file.write((char*)&text, sizeof(text));
    out_file.close();

    return 0;
}

File: source/pydgin.py
import regex as re
import os

"""
Single-function translation of Pydgin.
"""

def translate(inp_path: str, out_dir: str) -> str:
    with open(inp_path, 'r') as file:
        text = file.read()
    
    # Determine if header
    filename = out_dir + '/' + re.search(r'[^/\\.]+(?=\.(pdg)|(txt))', inp_path).group()
    if text[:8] == '@header':
        text = text[8:]
        filename += '.hpp'
    else:
        filename += '.cpp'

    # Try precompiled translators
    if os.path.exists('../bin/cpp_pdg.exe'):
        os.system('../bin/cpp_pdg.exe ' + inp_path + ' -o ' + filename)
    elif os.path.exists('../bin/cpp_pdg'):
        os.system('../bin/cpp_pdg ' + inp_path + ' -o  ' + filename)
    else:
        print('NO PRECOMPILED TRANSLATOR EXISTS.')
        print('Full Pydgin setup can be run with \'make setup\' in the pydgin folder.')
    
    if os.path.exists(filename):
        print('Successfull precompiled translator call')
        return filename
    else:
        print('FAILED PRECOMPILED TRANSLATOR CALL')
    
    text = re.sub(r'::', r' get ', text)
    text = re.sub(r'(?<![>:])\n', r';\n', text)
    text = re.sub(r'    ', r'\t', text)
    text = re.sub(r'private', r'\tprivate', text)
    text = re.sub(r'public', r'\tpublic', text)
    text = re.sub(r'protected', r'\tprotected', text)

    # Fix loop parethesizing
    loop_subs = {
        r'while (?=[^\(])': r'while (',
        r'for (?=[^\(])': r'for (',
        r'if (?=[^\(])': r'if (',
        r'else if (?=[^\(])': r'else if ('
    }
    for pat in loop_subs:
        text = re.sub(pat, loop_subs[pat], text)
    del loop_subs;

    # Brackets, end parenthesis, for loop semicoloning
    temp = re.split(r'\n', text)
    temp = [i + '\n' for i in temp]
    text = ''

    # Do tab replacement
    tab_count_prev = 0
    for line_num, line in enumerate(temp):
        # Get current tab count
        if len(line) == 0:
            continue

        # Fix for loop semicoloning
        if re.search(r'(?<!\w)for ', line):
            line = re.sub(r',', r';', line)

        # Balance loop parenthesis
        if re.search(r'(for )|(while )|(if )|(else if )', line):
            opens = 0
            prefix = ""
            for i, char in enumerate(line):
                if char == '(':
                    opens += 1
                if char == ')':
                    opens -= 1
                prefix += char
                if i + 1 == len(line) or line[i + 1] == ':':
                    for j in range(opens):
                        prefix += ')'
                    prefix += line[i + 1:]
                    break
            temp[line_num] = prefix
            
        line = temp[line_num]

        # Balance tabs with brackets
        tab_count_cur = 0
        while line[tab_count_cur] == '\t':
            tab_count_cur += 1

        # Compare tabs and adjust brackets
        if tab_count_cur > tab_count_prev:
            for i in range(tab_count_prev, tab_count_cur):
                text = text[:-1]
                text += ' {\n'
        elif tab_count_cur < tab_count_prev:
            for i in range(tab_count_prev, tab_count_cur, -1):
                text += ('\t' * (i - 1)) + '}' + '\n'

        # Reappend line to text
        text += line
        tab_count_prev = tab_count_cur

    # Do RegEx replacements
    replace = {
        r'(?<=\n).*#.*\n': r'',
        r'@import': r'#include',
        r'(?<!(private)|(public)):(?!=[a-zA-Z0-9:])': r'',
        r'[\'"]{3}[^\'"]*[\'"]{3}': r'',
        r'\n;': r'',
        r' in ': r' : ',
        r'\); {': r') {',
        r'(?<=else);': r'',
        r'@': r'#'
    }
    for pat in replace:
        text = re.sub(pat, replace[pat], text)
    del replace

    # Fix struct/class semicoloning
    temp = re.split(r'\n', text)
    text = ''
    to_semicolon = -1
    for i, line in enumerate(temp):
        tabs = 0
        while tabs + 1 < len(line) and line[tabs] == '\t':
            tabs += 1
        if re.search(r'(struct )|(class )', line):
            to_semicolon = tabs
        elif to_semicolon == tabs:
            temp[i] += ';'
            to_semicolon = -1
    text = '\n'.join(temp)

    # Untab public/private/restricted
    text = re.sub(r'\tprivate', r'private', text)
    text = re.sub(r'\tpublic', r'public', text)
    text = re.sub(r'\tprotected', r'protected', text)

    # Trim trailing whitespace
    while text[-1] == ' ' or text[-1] == '\n':
        text = text[:-1]
    text += '\n'
    
    text = re.sub(r' get ', r'::', text)
    text = re.sub(r'\t', r'    ', text)

    # Write to output
    with open(filename, 'w') as file:
        file.write(text)

    return filename